@codex please fix

```
✓ scripts/tests/validateQuest.test.ts (2 tests) 32ms
 ✓ tests/license.test.ts (3 tests) 6ms
 ✓ tests/questGuidelinesEditorDoc.test.ts (1 test) 6ms
 ✓ tests/ubiQuestLeaderboardCopy.test.ts (1 test) 4ms
 ✓ frontend/tests/playwrightConfigHost.test.ts (1 test) 4ms
 ✓ tests/roadmap10xQuests.test.ts (1 test) 4ms
 ✓ tests/questGraphValidation.test.ts (2 tests) 214ms
 ✓ tests/solarBasicsIcons.test.ts (5 tests) 5ms
 ✓ tests/rocketsDocsContent.test.ts (1 test) 7ms
 ✓ scripts/tests/lintStagedItemValidation.test.ts (1 test) 6ms
 ✓ scripts/tests/lintStagedQuestValidation.test.ts (1 test) 7ms
 ✓ tests/rocketsDocsPromise.test.ts (1 test) 3ms
 ✓ tests/fsChecksInvalidEncoding.test.ts (1 test) 9ms
 ✓ tests/roadmapSuborbitalLaunch.test.ts (1 test) 4ms
 ✓ tests/roadmapCustomQuests.test.ts (1 test) 5ms
 ✓ tests/changelogNotes.test.ts (1 test) 3ms
 ✓ tests/testingDocsPromises.test.ts (1 test) 3ms
 ✓ frontend/src/components/__tests__/compactItemListHelpers.spec.ts (1 test) 3ms
 ✓ tests/k3sChecklistComplete.test.ts (1 test) 4ms
 ✓ tests/docsFAQ.test.ts (1 test) 4ms
 ✓ tests/roadmapPromises.test.ts (1 test) 4ms
 ✓ tests/solarDocWind.test.ts (1 test) 4ms
 ✓ tests/docsPromptsOutages.test.ts (2 tests) 3ms
 ✓ tests/menu-active.test.ts (1 test) 4ms
 ✓ tests/sanitization.test.ts (1 test) 23ms
 ✓ tests/metricsFallback.test.ts (1 test) 43ms
 ✓ tests/itemSelectorTouch.test.ts (2 tests) 4ms
 ✓ tests/process-form-mobile.test.ts (1 test) 3ms
 ✓ tests/questCountIncrease.test.ts (1 test) 18ms
 ✓ tests/item-form-mobile.test.ts (1 test) 6ms
 ✓ tests/quest-form-mobile.test.ts (1 test) 4ms
 ✓ tests/titlesDoc.test.ts (1 test) 3ms
 ✓ tests/menu-leaderboard.test.ts (1 test) 6ms
 ✓ tests/togglePreviewId.test.ts (3 tests) 8ms
 ✓ scripts/tests/questQuality.test.ts (1 test) 5ms
 ✓ tests/npmTestCiScript.test.ts (1 test) 4ms
 ✓ tests/menu-settings.test.ts (1 test) 5ms
 ✓ tests/achievementsCatalog.test.ts (1 test) 5ms
 ✓ tests/mergePlanChecklist.test.ts (1 test) 6ms
 ✓ tests/menu-stats.test.ts (1 test) 5ms
 ✓ tests/profileTitles.test.ts (1 test) 4ms
 ✓ frontend/tests/menu.test.ts (1 test) 4ms
 ✓ frontend/tests/duration.test.ts (2 tests) 4ms
 ✓ tests/menuTitles.test.ts (1 test) 5ms
 ✓ tests/pnpmfile.test.ts (1 test) 5ms
 ✓ tests/domExceptionGlobal.test.ts (1 test) 4ms
 ✓ tests/domexception-global.test.ts (1 test) 3ms
 ✓ tests/domexception-globals.test.ts (1 test) 5ms
 ✓ tests/sysadminQuest.test.ts (1 test) 5ms
 ✓ tests/shopIndex.test.ts (1 test) 8ms
 ✓ tests/eslintPluginPresence.test.ts (1 test) 4ms
 ❯ scripts/tests/prettierFormatting.test.ts (1 test | 1 failed) 25110ms
   × Prettier formatting > frontend sources are formatted 25108ms
     → npm run format:check failed.

> dspace@3.0.0 format:check
> prettier --config .prettierrc --check "**/*.{js,ts,svelte,json,md}"

Checking formatting...
[warn] src/components/__tests__/LegacySaveUpgrade.spec.ts
[warn] src/pages/chat/svelte/OpenAIChat.svelte
[warn] tests/ensure-playwright-browsers.test.ts
[warn] Code style issues found in 3 files. Run Prettier with --write to fix.

 ✓ frontend/tests/gameStateFlush.test.ts (1 test) 4ms
 ✓ tests/checkoutFetchDepth.test.ts (1 test) 3ms
 ✓ tests/npmTestRunsAllSuites.test.ts (1 test) 4ms
 ✓ tests/astroConfigLogLevel.test.ts (1 test) 3ms
 ✓ tests/prismSvelte.test.ts (1 test) 15ms
 ✓ scripts/tests/dbBenchmarkScript.test.ts (1 test) 3ms
 ✓ tests/progressionBalance.test.ts (1 test) 33ms
 ✓ tests/domException.test.ts (1 test) 4ms
 ✓ frontend/tests/sanity.test.ts (1 test) 3ms
 ✓ tests/indexeddb.server.test.ts (1 test) 259ms

⎯⎯⎯⎯⎯⎯⎯ Failed Tests 2 ⎯⎯⎯⎯⎯⎯⎯

 FAIL  tests/qaDocsLinkFreshness.test.ts > QA docs test link freshness > v3.md links target the right test lines
AssertionError: Expected ../../frontend/src/components/__tests__/LegacySaveUpgrade.spec.ts#L129-L139 to include "shows v2 parse warnings when localStorage contains invalid JSON" (from v3.md:484): expected false to be true // Object.is equality

- Expected
+ Received

- true
+ false

 ❯ tests/qaDocsLinkFreshness.test.ts:62:11
     60|             doc
     61|           )}:${docLine})`
     62|         ).toBe(true);
       |           ^
     63|       }
     64|     });

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/2]⎯

 FAIL  scripts/tests/prettierFormatting.test.ts > Prettier formatting > frontend sources are formatted
Error: npm run format:check failed.

> dspace@3.0.0 format:check
> prettier --config .prettierrc --check "**/*.{js,ts,svelte,json,md}"

Checking formatting...
[warn] src/components/__tests__/LegacySaveUpgrade.spec.ts
[warn] src/pages/chat/svelte/OpenAIChat.svelte
[warn] tests/ensure-playwright-browsers.test.ts
[warn] Code style issues found in 3 files. Run Prettier with --write to fix.

 ❯ scripts/tests/prettierFormatting.test.ts:17:13
     15|       const stdout = error?.stdout?.toString?.() ?? '';

     16|       const stderr = error?.stderr?.toString?.() ?? '';
     17|       throw new Error(`npm run format:check failed.\n${stdout}${stderr…
       |             ^
     18|     }
     19|   }, 240000);

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[2/2]⎯

 Test Files  2 failed | 289 passed (291)
      Tests  2 failed | 1036 passed (1038)
   Start at  01:39:12
   Duration  153.15s (transform 9.61s, setup 10.51s, collect 43.51s, tests 110.20s, environment 181.44s, prepare 33.56s)


Error: AssertionError: Expected ../../frontend/src/components/__tests__/LegacySaveUpgrade.spec.ts#L129-L139 to include "shows v2 parse warnings when localStorage contains invalid JSON" (from v3.md:484): expected false to be true // Object.is equality

- Expected
+ Received

- true
+ false

 ❯ tests/qaDocsLinkFreshness.test.ts:62:11



Error: Error: npm run format:check failed.

> dspace@3.0.0 format:check
> prettier --config .prettierrc --check "**/*.{js,ts,svelte,json,md}"

Checking formatting...
[warn] src/components/__tests__/LegacySaveUpgrade.spec.ts
[warn] src/pages/chat/svelte/OpenAIChat.svelte
[warn] tests/ensure-playwright-browsers.test.ts
[warn] Code style issues found in 3 files. Run Prettier with --write to fix.

 ❯ scripts/tests/prettierFormatting.test.ts:17:13


 ELIFECYCLE  Command failed with exit code 1.
Error: Process completed with exit code 1.
```

new codex task, not a r/e/v/i/e/w task

--- a/docs/qa/v3.md
+++ b/docs/qa/v3.md
@@ -481,7 +481,7 @@ Local-mode validation status (March 20, 2026):
 - [x] Before launch, confirm there is at least one way to detect migration failures quickly (e.g.,
       QA console error review, or a logged event if available).
       ([<backups.md:Storage locations and QA reset notes (inline warnings + console)>](../../frontend/src/pages/docs/md/backups.md#storage-locations-and-qa-reset-notes),
-      [<LegacySaveUpgrade.spec.ts:shows v2 parse warnings when localStorage contains invalid JSON>](../../frontend/src/components/__tests__/LegacySaveUpgrade.spec.ts#L129-L139))
+      [<LegacySaveUpgrade.spec.ts:shows v2 parse warnings when localStorage contains invalid JSON>](../../frontend/src/components/__tests__/LegacySaveUpgrade.spec.ts#L141-L151))
 
 ### 3.3 External links (open + correct destination)
 

--- a/frontend/src/components/__tests__/LegacySaveUpgrade.spec.ts
+++ b/frontend/src/components/__tests__/LegacySaveUpgrade.spec.ts
@@ -27,10 +27,7 @@ describe('LegacySaveUpgrade', () => {
         const originalConsoleError = console.error;
         consoleErrorMock = vi.spyOn(console, 'error').mockImplementation((...args: unknown[]) => {
             const [first] = args;
-            if (
-                typeof first === 'string' &&
-                first.startsWith('Error reading from localStorage:')
-            ) {
+            if (typeof first === 'string' && first.startsWith('Error reading from localStorage:')) {
                 return;
             }
             originalConsoleError(...args);

--- a/frontend/src/pages/chat/svelte/OpenAIChat.svelte
+++ b/frontend/src/pages/chat/svelte/OpenAIChat.svelte
@@ -365,9 +365,7 @@
         const resolvedDocsRagSourceRef = docsMeta?.sourceRef ?? 'unknown';
         const resolvedDocsRagGeneratedAt = docsMeta?.generatedAt ?? 'unknown';
         const resolvedDocsRagHost =
-            typeof window !== 'undefined' && window.location
-                ? window.location.host
-                : 'unknown';
+            typeof window !== 'undefined' && window.location ? window.location.host : 'unknown';
         const resolvedHostEnvName = deriveEnvNameFromHostname(resolvedDocsRagHost);
         const allowDocsFallback = resolvedHostEnvName === 'dev';
         const appShaInfo = allowDocsFallback

--- a/frontend/src/pages/chat/svelte/OpenAIChat.svelte
+++ b/frontend/src/pages/chat/svelte/OpenAIChat.svelte
@@ -41,6 +41,7 @@
         getDocsRagMismatchWarning,
     } from '../../../utils/docsRag.js';
     import { copyToClipboard } from '../../../utils/copyToClipboard.js';
+    import { isBrowser } from '../../../utils/ssr.js';
     import Message from './Message.svelte';
     import Spinner from '../../../components/svelte/Spinner.svelte';
 
@@ -76,6 +77,7 @@
     let debugOverride = false;
     let currentSettings = { showChatDebugPayload: false };
     let lastShowChatDebugPayload;
+    let componentDestroyed = false;
     let playerStateSummary = {
         included: false,
         questsFinishedCount: 0,
@@ -413,6 +415,10 @@
         docsRagEnvWarning = buildDocsEnvMismatchWarning(docsRagHost, docsRagEnvName);
         appGitSha = resolvedAppShaInfo.sha;
         playerStateSummary = getPlayerStateSummary(currentState);
+        if (componentDestroyed || !isBrowser) {
+            return;
+        }
+
         syncSaveSnapshotHintDismissed();
         saveSnapshotHintFocusListener = () => syncSaveSnapshotHintDismissed();
         window.addEventListener('focus', saveSnapshotHintFocusListener);
@@ -437,11 +443,12 @@
     });
 
     onDestroy(() => {
+        componentDestroyed = true;
         settingsUnsubscribe?.();
-        if (saveSnapshotHintFocusListener) {
+        if (isBrowser && saveSnapshotHintFocusListener) {
             window.removeEventListener('focus', saveSnapshotHintFocusListener);
         }
-        if (promptDebugLinkListener) {
+        if (isBrowser && promptDebugLinkListener) {
             window.removeEventListener('hashchange', promptDebugLinkListener);
             window.removeEventListener('popstate', promptDebugLinkListener);
         }

--- a/frontend/tests/ensure-playwright-browsers.test.ts
+++ b/frontend/tests/ensure-playwright-browsers.test.ts
@@ -218,7 +218,9 @@ describe('ensurePlaywrightBrowsers', () => {
             const [first] = args;
             if (
                 typeof first === 'string' &&
-                (first.startsWith('Playwright chromium executable is still missing after installation.') ||
+                (first.startsWith(
+                    'Playwright chromium executable is still missing after installation.'
+                ) ||
                     first.startsWith(
                         'PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD=1 but Playwright chromium browser is missing.'
                     ))

